File: backend/src/main/java/com/backend/dtos/BaseSummaryDTO.java
package com.backend.dtos;

import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.PositiveOrZero;

import static com.backend.utils.ValidationPatterns.TIME_PATTERN;

/**
 * A summary of key health metrics.
 * <p>
 * Includes heart rate, resting heart rate, calorie burn, weight,
 * hydration, stress, respiration rate, sleep duration, and SpO2 levels.
 * Each field has predefined validation constraints to ensure correct data.
 * </p>
 *
 * This DTO is built from the corresponding model record {@code com.backend.models.BaseSummary}.
 */
public record BaseSummaryDTO(
    // ------ Heart Rate ----- //
    @Min(20) @Max(250)
    Integer hrMin,    // Min heart rate (20-250 bpm)
    @Min(20) @Max(250)
    Integer hrMax,    // Max heart rate (20-250 bpm)
    @Min(20) @Max(250)
    Integer hrAvg,    // Avg heart rate (20-250 bpm)
    @Min(20) @Max(150)
    Integer rhrMin,   // Min resting heart rate (20-150 bpm)
    @Min(20) @Max(150)
    Integer rhrMax,   // Max resting heart rate (20-150 bpm)
    @Min(20) @Max(150)
    Integer rhrAvg,   // Avg resting heart rate (20-150 bpm)
    @Min(20) @Max(150)
    Integer inactiveHrMin, // Min inactive heart rate
    @Min(20) @Max(150)
    Integer inactiveHrMax, // Max inactive heart rate
    @Min(20) @Max(150)
    Integer inactiveHrAvg, // Avg inactive heart rate

    // ------ Calories ----- //
    @PositiveOrZero
    Integer caloriesAvg,    // Avg daily calories burned (≥0)
    @PositiveOrZero
    Integer caloriesGoal,   // Target calorie burn goal (≥0)
    @PositiveOrZero
    Integer caloriesBmrAvg, // Avg Basal Metabolic Rate (BMR) (≥0)
    @PositiveOrZero
    Integer caloriesConsumedAvg, // Avg daily calories consumed (≥0)
    @PositiveOrZero
    Integer caloriesActiveAvg,  // Avg active calories burned (≥0)
    @PositiveOrZero
    Integer activitiesCalories, // Total calories burned in activities (≥0)

    // ------ Weight ----- //
    @Min(0) @Max(300)
    Double weightMin,   // Min weight recorded (0-300 kg)
    @Min(0) @Max(300)
    Double weightMax,   // Max weight recorded (0-300 kg)
    @Min(0) @Max(300)
    Double weightAvg,   // Avg weight recorded (0-300 kg)

    // ------ Hydration ----- //
    @PositiveOrZero
    Integer hydrationGoal,   // Target hydration intake (ml, ≥0)
    @PositiveOrZero
    Integer hydrationIntake, // Total hydration intake (ml, ≥0)
    @PositiveOrZero
    Integer hydrationAvg,    // Avg daily hydration intake (ml, ≥0)
    @PositiveOrZero
    Integer sweatLoss,       // Total sweat loss (ml, ≥0)
    @PositiveOrZero
    Integer sweatLossAvg,    // Avg sweat loss (ml, ≥0)

    // ------ Stress & Body Battery ----- //
    @Min(0) @Max(100)
    Integer bbMin,    // Min body battery (0-100)
    @Min(0) @Max(100)
    Integer bbMax,    // Max body battery (0-100)
    @Min(0) @Max(100)
    Integer stressAvg, // Avg stress level (0-100)

    // ------ Respiration & SPO ----- //
    @Min(5) @Max(50)
    Integer rrMin,    // Min respiration rate (5-50 bpm)
    @Min(5) @Max(50)
    Integer rrMax,    // Max respiration rate (5-50 bpm)
    @Min(5) @Max(50)
    Integer rrWakingAvg,  // Avg waking respiration rate (5-50 bpm)
    @Min(70) @Max(100)
    Integer spo2Min, // Min SpO2 level (70-100%)
    @Min(70) @Max(100)
    Integer spo2Avg, // Avg SpO2 level (70-100%)

    // ------ Sleep ----- //
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String sleepMin,   // Min sleep duration (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String sleepMax,   // Max sleep duration (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String sleepAvg,   // Avg sleep duration (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String remSleepMin, // Min REM sleep duration (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String remSleepMax, // Max REM sleep duration (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String remSleepAvg, // Avg REM sleep duration (HH:MM:SS format)

    // ------ Steps & Floors ----- //
    @PositiveOrZero
    Integer stepsGoal,  // Target step count (≥0)
    @PositiveOrZero
    Integer steps,      // Total steps taken (≥0)
    @PositiveOrZero
    Integer floorsGoal, // Target floors climbed (≥0)
    @PositiveOrZero
    Integer floors,     // Total floors climbed (≥0)

    // ------ Activities ----- //
    @PositiveOrZero
    Integer activities, // Total logged activities (≥0)
    @PositiveOrZero
    Double activitiesDistance, // Total distance traveled (km, ≥0)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String intensityTimeGoal, // Total intensity time goal (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String intensityTime, // Total intensity minutes (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String moderateActivityTime, // Total time in moderate activity (HH:MM:SS format)
    @Pattern(regexp = TIME_PATTERN, message = "Invalid time format (HH:MM:SS)")
    String vigorousActivityTime // Total time in vigorous activity (HH:MM:SS format)
) {
    /**
     * Converts a {@link com.backend.models.BaseSummary} record to a {@link BaseSummaryDTO}.
     *
     * @param model the BaseSummary model from the persistence layer.
     * @return a new BaseSummaryDTO populated with values from the model.
     */
    public static BaseSummaryDTO fromModel(com.backend.models.BaseSummary model) {
        return new BaseSummaryDTO(
            model.hrMin(),
            model.hrMax(),
            model.hrAvg(),
            model.rhrMin(),
            model.rhrMax(),
            model.rhrAvg(),
            model.inactiveHrMin(),
            model.inactiveHrMax(),
            model.inactiveHrAvg(),
            model.caloriesAvg(),
            model.caloriesGoal(),
            model.caloriesBmrAvg(),
            model.caloriesConsumedAvg(),
            model.caloriesActiveAvg(),
            model.activitiesCalories(),
            model.weightMin(),
            model.weightMax(),
            model.weightAvg(),
            model.hydrationGoal(),
            model.hydrationIntake(),
            model.hydrationAvg(),
            model.sweatLoss(),
            model.sweatLossAvg(),
            model.bbMin(),
            model.bbMax(),
            model.stressAvg(),
            model.rrMin(),
            model.rrMax(),
            model.rrWakingAvg(),
            model.spo2Min(),
            model.spo2Avg(),
            model.sleepMin(),
            model.sleepMax(),
            model.sleepAvg(),
            model.remSleepMin(),
            model.remSleepMax(),
            model.remSleepAvg(),
            model.stepsGoal(),
            model.steps(),
            model.floorsGoal(),
            model.floors(),
            model.activities(),
            model.activitiesDistance(),
            model.intensityTimeGoal(),
            model.intensityTime(),
            model.moderateActivityTime(),
            model.vigorousActivityTime()
        );
    }
}

File: frontend/src/types/BaseSummaryDTO.ts
/**
 * A summary of key health metrics.
 * This type is built based on the backend BaseSummaryDTO.
 */
export type BaseSummaryDTO = {
  // ------ Heart Rate ----- //
  hrMin: number;           // Min heart rate (20-250 bpm)
  hrMax: number;           // Max heart rate (20-250 bpm)
  hrAvg: number;           // Avg heart rate (20-250 bpm)
  rhrMin: number;          // Min resting heart rate (20-150 bpm)
  rhrMax: number;          // Max resting heart rate (20-150 bpm)
  rhrAvg: number;          // Avg resting heart rate (20-150 bpm)
  inactiveHrMin: number;   // Min inactive heart rate
  inactiveHrMax: number;   // Max inactive heart rate
  inactiveHrAvg: number;   // Avg inactive heart rate

  // ------ Calories ----- //
  caloriesAvg: number;         // Avg daily calories burned (≥0)
  caloriesGoal: number;        // Target calorie burn goal (≥0)
  caloriesBmrAvg: number;      // Avg Basal Metabolic Rate (BMR) (≥0)
  caloriesConsumedAvg: number; // Avg daily calories consumed (≥0)
  caloriesActiveAvg: number;   // Avg active calories burned (≥0)
  activitiesCalories: number;  // Total calories burned in activities (≥0)

  // ------ Weight ----- //
  weightMin: number;       // Min weight recorded (0-300 kg)
  weightMax: number;       // Max weight recorded (0-300 kg)
  weightAvg: number;       // Avg weight recorded (0-300 kg)

  // ------ Hydration ----- //
  hydrationGoal: number;    // Target hydration intake (ml, ≥0)
  hydrationIntake: number;  // Total hydration intake (ml, ≥0)
  hydrationAvg: number;     // Avg daily hydration intake (ml, ≥0)
  sweatLoss: number;        // Total sweat loss (ml, ≥0)
  sweatLossAvg: number;     // Avg sweat loss (ml, ≥0)

  // ------ Stress & Body Battery ----- //
  bbMin: number;    // Min body battery (0-100)
  bbMax: number;    // Max body battery (0-100)
  stressAvg: number; // Avg stress level (0-100)

  // ------ Respiration & SPO2 ----- //
  rrMin: number;       // Min respiration rate (5-50 bpm)
  rrMax: number;       // Max respiration rate (5-50 bpm)
  rrWakingAvg: number; // Avg waking respiration rate (5-50 bpm)
  spo2Min: number;     // Min SpO2 level (70-100%)
  spo2Avg: number;     // Avg SpO2 level (70-100%)

  // ------ Sleep ----- //
  // Expected format: "HH:MM:SS"
  sleepMin: string;   // Min sleep duration
  sleepMax: string;   // Max sleep duration
  sleepAvg: string;   // Avg sleep duration
  remSleepMin: string; // Min REM sleep duration
  remSleepMax: string; // Max REM sleep duration
  remSleepAvg: string; // Avg REM sleep duration

  // ------ Steps & Floors ----- //
  stepsGoal: number;  // Target step count (≥0)
  steps: number;      // Total steps taken (≥0)
  floorsGoal: number; // Target floors climbed (≥0)
  floors: number;     // Total floors climbed (≥0)

  // ------ Activities ----- //
  activities: number;         // Total logged activities (≥0)
  activitiesDistance: number; // Total distance traveled (km, ≥0)
  // Expected format: "HH:MM:SS"
  intensityTimeGoal: string;  // Total intensity time goal
  intensityTime: string;      // Total intensity minutes
  moderateActivityTime: string; // Total moderate activity time
  vigorousActivityTime: string; // Total vigorous activity time
}
